app/backend/src/couchers/jobs/worker.py

```diff
@@ -0,0 +1,177 @@
+                job.next_attempt_after += timedelta(seconds=15 * (2 ** job.try_count))
+                logger.info(f"Job #{job.id} error on try number {job.try_count}, next try at {job.next_attempt_after}")
+            # add some info for debugging
+            job.failure_info = traceback.format_exc()
+
+        # exiting ctx manager commits and releases the row lock
+
+
+def service_jobs():
+    """
+    Don't run many of these simultaneously: it'll work but it'll conflict and cause a lot of missed job processing.
+    """
+    # multiprocessing uses fork() which in turn copies file descriptors, so the engine may have connections in its pool
+    # that we don't want to reuse. This is the SQLALchemy-recommended way of clearing the connection pool in this thread
+    get_engine().dispose()
+
+    MAX_WORKERS = 8
+    with futures.ThreadPoolExecutor(max_workers=MAX_WORKERS) as executor:
+        jobs = []
+        while True:
+            # jobs should be list of currently running jobs
+            jobs = [job for job in jobs if not job.done()]
+            idle_workers = MAX_WORKERS - len(jobs)
+            if idle_workers == 0:
+                # if we're at max capacity, then wait for one job to finish
+                futures.wait(jobs, return_when=futures.FIRST_COMPLETED)
+
+                # loop back to start, it's cleaner
+                continue
+
+            with session_scope() as session:
+                subquery = (
+                    session.query(BackgroundJob.id)
+                    .filter(BackgroundJob.ready_for_retry)
+                    .limit(idle_workers)
+                    .with_for_update(skip_locked=True)
+                    .subquery()
+                )
+                job_count = session.query(subquery).count()
+                # end the transaction and don't keep it open while sleeping, etc
+
+            if job_count > 0:
+                for _ in range(min(job_count, idle_workers)):
+                    jobs.append(executor.submit(process_job))
+            else:
+                # no jobs to work on, don't hammer the DB, we can wait a second
+                sleep(1)
+
+
+def _run_job_and_schedule(sched, schedule_id):
+    job_type, frequency = SCHEDULE[schedule_id]
+    logger.info(f"Processing job of type {job_type}")
+    with session_scope() as session:
+        rjob = session.query(RepeatedJob).filter(RepeatedJob.job_type == job_type).one()
+        rjob.last_run = func.now()
+
+    # wake ourselves up after frequency
+    sched.enter(
+        frequency.total_seconds(),
+        1,
+        _run_job_and_schedule,
+        argument=(
+            sched,
+            schedule_id,
+        ),
+    )
+
+    # queue the job
+    queue_job(job_type, empty_pb2.Empty())
+
+
+def run_scheduler():
+    """
+    Schedules jobs according to schedule in .definitions
+    """
+    # multiprocessing uses fork() which in turn copies file descriptors, so the engine may have connections in its pool
```

Comment:
At the moment the app checks the db for a connection before basically doing anything, and fails fast if it can't ping the DB. (This happens in app.py). It's much easier to catch that way.

I think it's less error prone to keep these in here, and it took me quite a while to figure it out (hence the big comment...).

Reply:
We can remove this if we later refactor to fork in a shell script. Resolving.